doc: orphaned instance deletion (#1095)
longhorn/website: content/docs/1.9.0/advanced-resources/data-cleanup/orphaned-instance-cleanup.md

---
title: Orphaned Instance Cleanup
weight: 4
---

> Longhorn supports orphaned instance cleanup. Currently, Longhorn can identify and clean up the orphaned instances on each node.
> 
> ## Orphaned Instance Runtime
> 
> When a network outage occurs to some Longhorn node, it may contain multiple engine or replica runtime instances not tracked by the Longhorn System. The corresponding engine and replica CRs might be removed or rescheduled to another node during the outage. When the node comes back, the Longhorn system no longer tracks the corresponding runtime resources. These runtime resources, including the processes, are called orphaned instances. Orphaned instances continue to consume CPU and memory.
> 
> Longhorn supports the detection and cleanup of orphaned instance processes. It identifies the processes and gives a list of `orphan` resources that describe those processes. By default, Longhorn does not automatically delete `orphan` resources and their processes. Users can trigger the deletion of orphaned instances manually or have it done automatically.
> 
> ### Example
> 
> In the example, we will explain how to manage orphaned instances identified by Longhorn via `kubectl`.

Reviewer: ```suggestion
Longhorn can identify and clean up orphaned instance processes on each node.

## Orphaned Instance Runtime

When a network outage affects a Longhorn node, it may leave behind engine or replica runtime instances that are no longer tracked by the Longhorn system. The corresponding engine and replica CRs might be removed or rescheduled to another node during the outage. When the node comes back, the Longhorn system no longer tracks the corresponding runtime resources. These runtime resources, including the processes, are called orphaned instances. Orphaned instances continue to consume CPU and memory.

Longhorn supports the detection and cleanup of orphaned instance processes. It identifies the processes and gives a list of `orphan` resources that describe those processes. By default, Longhorn does not automatically delete `orphan` resources and their processes. Users can trigger the deletion of orphaned instances manually or have it done automatically.

### Example

The following example shows how to manage orphaned instances using `kubectl`.
```

    ```
    # kubectl -n longhorn-system get orphans.longhorn.io -l "longhorn.io/orphan-type in (engine-instance,replica-instance)"
    No resources found in longhorn-system namespace.
    ```
    The orphaned instances are deleted from instance manager.
    ```
    # kubectl -n longhorn-system describe instancemanager -l "longhorn.io/node=worker1"
    Name:         instance-manager-8ff396d6d3744979b32abafc6346781c
    Namespace:    longhorn-system
    Kind:         InstanceManager
    ...
    Status:
      Instance Replicas:
    ...

    # kubectl -n longhorn-system describe instancemanager -l "longhorn.io/node=worker2"
    Name:         instance-manager-b87f10b867cec1dca2b814f5e78bcc90
    Namespace:    longhorn-system
    Kind:         InstanceManager
    ...
    Status:
      Instance Engines:
    ...

    ```

    Additionally, one can delete all orphaned instances on the specified node by
    ```
    # kubectl -n longhorn-system delete orphan -l "longhorn.io/orphan-type in (engine-instance,replica-instance),longhornnode=<node name>"
    ```

#### Manage Orphaned Instances via Longhorn UI

> In the top navigation bar of the Longhorn UI, click `Setting > Orphaned Data > Instance Orphans`. Orphaned instances on each node are listed. One can delete the instances by `Operation > Delete`.
> 
> By default, Longhorn will not automatically delete the orphaned instances. One can enable the automatic deletion in `Setting > General > Orphan`.
> 
> ### Exception
> Longhorn will not create an `orphan` resource for an orphaned instance when
> - The orphaned engine or replica is rescheduled back to the node.
> - The orphaned engine or replica is in migrating, starting, or stopping.
> - Node is evicted.

Reviewer: ```suggestion
In the top navigation bar of the Longhorn UI, click `Setting > Orphaned Data > Instance Orphans`. Orphaned instances on each node are listed. You can delete instances by selecting `Operation > Delete`.

By default, Longhorn does not automatically delete orphaned instances.
To enable automatic deletion, go to `Setting > General > Orphan`.

### Exception
Longhorn does not create an orphan resource in the following scenarios:
- The orphaned engine or replica is rescheduled back to the node.
- The engine or replica is in a migrating, starting, or stopping state.
- The node is evicted.
```

---
title: Orphaned Instance Cleanup
weight: 4
---

> Longhorn can identify and clean up orphaned instance processes on each node.

Reviewer 2: ```suggestion
Longhorn can identify and clean up orphaned instances on each node.
```

## Orphaned Runtime Instance

> When a network outage affects a Longhorn node, it may leave behind engine or replica runtime instances that are no longer tracked by the Longhorn system. The corresponding engine and replica CRs might be removed or rescheduled to another node during the outage. When the node comes back, the Longhorn system no longer tracks the corresponding runtime resources. These runtime resources, including the processes, are called orphaned instances. Orphaned instances continue to consume CPU and memory.

Reviewer 2: ```suggestion
When a network outage affects a Longhorn node, it may leave behind engine or replica runtime instances that are no longer tracked by the Longhorn system. The corresponding engine and replica CRs might be removed or rescheduled to another node during the outage. When the node comes back, the Longhorn system no longer tracks the corresponding runtime instances. These runtime instances, such as the engine and replica processes for v1 volumes, are called orphaned instances. Orphaned instances continue to consume CPU and memory.
```

> Longhorn supports the detection and cleanup of orphaned instance processes. It identifies the processes and gives a list of `orphan` resources that describe those processes. By default, Longhorn does not automatically delete `orphan` resources and their processes. Users can trigger the deletion of orphaned instances manually or have it done automatically.

Reviewer 2: ```suggestion
Longhorn supports the detection and cleanup of orphaned instance. It identifies the instances and gives a list of `orphan` resources that describe those orphans. By default, Longhorn does not automatically delete `orphan` instances. Users can trigger the deletion of orphaned instances manually or have it done automatically.
```

             Resource Version:   0
             State:              running
             Target Port End:    0
             Target Port Start:  0
             Type:               replica
     ...
     ```
    - Orphaned engine instance on Node `worker2`
     ```
     # kubectl -n longhorn-system describe instancemanager -l "longhorn.io/node=worker2"
     Name:         instance-manager-b87f10b867cec1dca2b814f5e78bcc90
     Namespace:    longhorn-system
     Kind:         InstanceManager
     ...
     Status:
       Instance Engines:
         pvc-569e44c0-b352-4aca-bf14-2cf7a6cfe86f-e-0:
           Spec:
             Data Engine:  v1
             Name:         pvc-569e44c0-b352-4aca-bf14-2cf7a6cfe86f-e-0
           Status:
             Conditions:
               Filesystem Read Only:  false
             Endpoint:
             Error Msg:
             Listen:
             Port End:                10020
             Port Start:              10020
             Resource Version:        0
             State:                   running
             Target Port End:         10020
             Target Port Start:       10020
             Type:                    engine
     ...
     ```

> 2. Longhorn detects the orphaned instances and creates an `orphan` resources describing the resources.

Reviewer 2: ```suggestion
2. Longhorn detects the orphaned instances and creates an `orphan` resources describing the instances.
```

    ```
    # kubectl -n longhorn-system get orphan -l "longhorn.io/orphan-type in (engine-instance,replica-instance)"
    NAME                                                                      TYPE               NODE
    orphan-1807009489e50534c35c350e22680449c97deca4e5d3b72f4591976145f8bc41   engine-instance    worker2
    orphan-a91aa42ab5eda6b8b9fe1116d5b5f5673e5108d89be3db6fd18a275913463eef   replica-instance   worker1
    ```

3. One can list the `orphan` resources created by Longhorn system by `kubectl -n longhorn-system get orphan`.
    ```
    kubectl -n longhorn-system get orphan
    ```

> 4. Get the detailed information of one of the orphaned replica instance in `spec.parameters` by `kubcel -n longhorn-system get orphan <name>`.

Reviewer: ```suggestion
4. Get the detailed information of one of the orphaned replica instance in `spec.parameters` by `kubectl -n longhorn-system get orphan <name>`.
```

    # kubectl -n longhorn-system get orphans orphan-a91aa42ab5eda6b8b9fe1116d5b5f5673e5108d89be3db6fd18a275913463eef -o yaml
    apiVersion: longhorn.io/v1beta2
    kind: Orphan
    metadata:
    creationTimestamp: "2025-05-02T06:07:32Z"
    finalizers:
    - longhorn.io
    generation: 1
    labels:
        longhorn.io/component: orphan
        longhorn.io/managed-by: longhorn-manager
        longhorn.io/orphan-type: replica-instance
        longhornnode: worker1
        longhornreplica: pvc-569e44c0-b352-4aca-bf14-2cf7a6cfe86f-r-05660b73

    ......

    spec:
      dataEngine: v1
      nodeID: worker1
      orphanType: replica-instance
      parameters:
        InstanceManager: instance-manager-8ff396d6d3744979b32abafc6346781c
        InstanceName: pvc-569e44c0-b352-4aca-bf14-2cf7a6cfe86f-r-05660b73
    status:
    conditions:
    - lastProbeTime: ""
        lastTransitionTime: "2025-05-02T06:06:39Z"
        message: ""
        reason: running
        status: "True"
        type: InstanceExist
    - lastProbeTime: ""
        lastTransitionTime: "2025-05-02T06:06:39Z"
        message: ""
        reason: ""
        status: "False"
        type: Error
    ownerID: worker1
    ```

> 5. Get the detailed information of one of the orphaned engine instance in `spec.parameters` by `kubcel -n longhorn-system get orphan <name>`.

Reviewer: ```suggestion
5. Get the detailed information of one of the orphaned engine instance in `spec.parameters` by `kubectl -n longhorn-system get orphan <name>`.
```

    ```
    # kubectl -n longhorn-system get orphans orphan-1807009489e50534c35c350e22680449c97deca4e5d3b72f4591976145f8bc41 -o yaml
    apiVersion: longhorn.io/v1beta2
    kind: Orphan
    metadata:
    creationTimestamp: "2025-05-02T06:47:25Z"
    finalizers:
    - longhorn.io
    generation: 1
    labels:
        longhorn.io/component: orphan
        longhorn.io/managed-by: longhorn-manager
        longhorn.io/orphan-type: engine-instance
        longhornengine: pvc-569e44c0-b352-4aca-bf14-2cf7a6cfe86f-e-0
        longhornnode: worker2

    ......

    spec:
      dataEngine: v1
      nodeID: worker2
      orphanType: engine-instance
      parameters:
        InstanceManager: instance-manager-b87f10b867cec1dca2b814f5e78bcc90
        InstanceName: pvc-569e44c0-b352-4aca-bf14-2cf7a6cfe86f-e-0
    status:
    conditions:
    - lastProbeTime: ""
        lastTransitionTime: "2025-05-02T06:47:25Z"
        message: ""
        reason: running
        status: "True"
        type: InstanceExist
    - lastProbeTime: ""
        lastTransitionTime: "2025-05-02T06:47:25Z"
        message: ""
        reason: ""
        status: "False"
        type: Error
    ownerID: worker2
    ```

> 6. One can delete the `orphan` resource by `kubectl -n longhorn-system delete orphan <name>` and then the corresponding orphaned instance will be deleted.

Reviewer: ```suggestion
6. You can delete an `orphan` resource by running `kubectl -n longhorn-system delete orphan <name>`. The corresponding orphaned instance will also be removed.
```

    ```
    # kubectl -n longhorn-system delete orphan orphan-a91aa42ab5eda6b8b9fe1116d5b5f5673e5108d89be3db6fd18a275913463eef

    # kubectl -n longhorn-system get orphan -l "longhorn.io/orphan-type in (engine-instance,replica-instance)"
    NAME                                                                      TYPE               NODE
    orphan-1807009489e50534c35c350e22680449c97deca4e5d3b72f4591976145f8bc41   engine-instance    worker2
    ```

    The orphaned instance is deleted.
    ```
    # kubectl -n longhorn-system describe instancemanager -l "longhorn.io/node=worker1"
    Name:         instance-manager-8ff396d6d3744979b32abafc6346781c
    Namespace:    longhorn-system
    Kind:         InstanceManager
    ...
    Status:
      Instance Replicas:
    ...
    ```

> 7. By default, Longhorn will not automatically delete the orphaned instance. One can enable the automatic deletion by setting `orphan-resource-auto-deletion`.

Reviewer: ```suggestion
7. By default, Longhorn does not automatically delete orphaned instances. You can enable automatic deletion by configuring the `orphan-resource-auto-deletion` setting.
```

    NAME                            VALUE                            APPLIED   AGE
    orphan-resource-auto-deletion   engineInstance;replicaInstance   true      45h
    ```

8. After enabling the automatic deletion and wait for a while, the `orphan` resources and processes are deleted automatically.
    ```
    # kubectl -n longhorn-system get orphans.longhorn.io -l "longhorn.io/orphan-type in (engine-instance,replica-instance)"
    No resources found in longhorn-system namespace.
    ```
    The orphaned instances are deleted from instance manager.
    ```
    # kubectl -n longhorn-system describe instancemanager -l "longhorn.io/node=worker1"
    Name:         instance-manager-8ff396d6d3744979b32abafc6346781c
    Namespace:    longhorn-system
    Kind:         InstanceManager
    ...
    Status:
      Instance Replicas:
    ...

    # kubectl -n longhorn-system describe instancemanager -l "longhorn.io/node=worker2"
    Name:         instance-manager-b87f10b867cec1dca2b814f5e78bcc90
    Namespace:    longhorn-system
    Kind:         InstanceManager
    ...
    Status:
      Instance Engines:
    ...

    ```

>     Additionally, one can delete all orphaned instances on the specified node by

Reviewer: ```suggestion
    Additionally, you can delete all orphaned instances on the specified node by running:
```

    ```
    # kubectl -n longhorn-system delete orphan -l "longhorn.io/orphan-type in (engine-instance,replica-instance),longhornnode=<node name>"
    ```

#### Manage Orphaned Instances via Longhorn UI

> In the top navigation bar of the Longhorn UI, click `Setting > Orphaned Data > Instance Orphans`. Orphaned instances on each node are listed. You can delete instances by selecting `Operation > Delete`.

Reviewer: ```suggestion
In the top navigation bar of the Longhorn UI, go to `Setting > Orphaned Data > Instance Orphans`. Orphaned instances on each node are listed. You can delete instances by selecting `Operation > Delete`.
```
